### DISCUSSION TITLE
Best Practice Guide for Relationship Objects

### DISCUSSION THREAD

**Discussion by maphe** [mikro-orm/mikro-orm#1411] [Q&A]
**Is your feature request related to a problem? Please describe.**
I'm trying to wrap my head around references in mikro orm, mix and matching different documentation pages as well as some far memories of using Doctrine... It remains unclear what is the "best" way to use relation objects, or which solution is the best suited for different scenarios.

**Describe the solution you'd like**
Following documentations kinda seem to conflict with each other, or serve different purposes:
- https://mikro-orm.io/docs/entity-references
- https://mikro-orm.io/docs/nested-populate

I would like some better documentation confronting both ways of dealing with relation objects. It's very easy to end up with a relation that's not initialized when the code is split up in many reusable functions dealing with different entities. It would be nice to have a guide that covers best practices for:
1. when to define relations as direct object or wrapped reference
2. how to ensure references are loaded in either way

**Describe alternatives you've considered**
As far as I understand, there are 2 ways to represent relations in an entity (eg. from a `Book` entity):
#### Direct
```
@ManyToOne()
author!: Author;
```
#### Wrapped
```
@ManyToOne()
author!: Reference<Author>;
```
which are both valid but must be handled in very different ways, although in both cases mikro-orm will internally work with references.

Now let's say I have a function that takes a book but requires the author to be loaded, I imagine respective solutions to be:
#### Direct
Force the relation to be populated ahead via join and add a safe guard first thing in the function.
```
function doSomething(book: Book) {
  if (!book.author.initialized) throw new Error();
  //...
}
```
Although this doesn't seem to be available, __initialized being private.

#### Wrapped
Ensure the reference is loaded.
```
function doSomething(book: Book) {
  const author = await book.author.load();
  //...
}
```
This seems to be safer but forces a lot of load/get()/wrap/unwrap shenanigans making the code a bit less readable.

**Additional context**
If I remember correctly my old Doctrine days, Doctrine generates Proxies and does everything under the hood, as if a property for a a non-initialized object is accessed, the proxy will load before returning the value. Is that even doable in ts/js for a future version of mikro-orm? Or does the language forces a more explicit/manual reference handling?

**Comment by B4nan** [comment]
The doctrine way is not doable, as you need async to do the conditional loading. Its actually the references way, which forces you to await things - verbose, but typesafe. 

The two pages you linked are not really about two approaches, one is about how you can define a relation (is more typesafe way), the other is about how to load them. You can define any relation as a reference, populating them is done in the very same way. 

Lastly, you can check whether an entity is loaded via `wrap(e).isInitialized()`. 

Ideally we should have the information whether a reference is loaded or not in the return type - I already tried to implement this in #214, but its too performance heavy. TS is unfortunately not very well suited for this, not without going all in for codegen (or maybe I just need to learn more of its magic, lol).

**Comment by maphe** [comment]
Thanks for the tip on `wrap(e).isInitialized()`.

I understand the 2 doc pages address different topics but the example in Smart Nested is only relevant if the relation is defined without the reference wrapper, which is why I was confronting them.

Also the way relations are defined (with or without the wrapper) has a significant impact on how the rest of the application has to be written, and it's pretty tedious to change from one to the other. I still think the documentation in this regard could be improved (I might just push a PR at some point as I'm getting more used to the pros and cons of each).

From your experience, do you have a preferred way over the other one? Looks like both require some verbosity dealing with relations anyway; I'm too early in my dev process to anticipate which one will be better down the road.

**Comment by B4nan** [comment]
> I understand the 2 doc pages address different topics but the example in Smart Nested is only relevant if the relation is defined without the reference wrapper, which is why I was confronting them.

Not really, the only difference is that you need to `unwrap()` everything (as you as a developer are the one who is responsible to know what is populated and what not - that is the missing bit that is hard to automate). In fact the `Reference` wrapper is purely about TS. There are "references" even if you don't use it, a reference is an entity that has only a PK. The `Reference` wrapper helps just with making this mandatory and typesafe. 

> Also the way relations are defined (with or without the wrapper) has a significant impact on how the rest of the application has to be written, and it's pretty tedious to change from one to the other.

Again, not really, everything works the same, using `Reference` wrapper just forces you to make the code typesafe by awaiting things. 

> From your experience, do you have a preferred way over the other one? Looks like both require some verbosity dealing with relations anyway; I'm too early in my dev process to anticipate which one will be better down the road.

Personally I am not using the reference wrapper at all, as it is quite tedious (as you need to await everything all the way). The `Loaded` type and automatic mapping to it based on the `populate` param (mentioned in #214) would help a lot, but as I said, it is really performance heavy.

**Comment by maphe** [comment]
>> Also the way relations are defined (with or without the wrapper) has a significant impact on how the rest of the application has to be written, and it's pretty tedious to change from one to the other.

> Again, not really, everything works the same, using Reference wrapper just forces you to make the code typesafe by awaiting things.

It may work the same internally but still you have to deal with objects differently; either wrapping entities to access the reference API (load, isInitialized...), or unwrapping the references to access the entities.

I think I get a better sense of it now. I'm probably gonna go with some kind of safe guard like:
```
async doSomething(book: Book) {
  if (!wrap(book.author).isInitialized()) await wrap(book.author).init();
  // or
  if (!wrap(book.author).isInitialized()) throw 'Author has to be loaded';
}
```
This seems to be the best compromise, I agree dealing with reference wrappers all over the place is a bit of an overhead.

Thank for the help and the quick responses. That's super helpful !

**Comment by kinthaiofficial** [comment]
MikroORM's relationship handling is one of its strongest points — here's best practice for the common patterns:

**1. Lazy loading (default) — be explicit about initialization**:
```typescript
@Entity()
class Author {
  @PrimaryKey() id: number;
  @Property() name: string;
  
  // Lazy by default — must call author.books.init() or use populate
  @OneToMany(() => Book, book => book.author)
  books = new Collection<Book>(this);
}

// Correct: always populate explicitly
const author = await em.findOne(Author, 1, { populate: ['books'] });
author.books.getItems();  // Safe — already loaded

// Wrong: accessing without populate
const author = await em.findOne(Author, 1);
author.books.getItems();  // Throws: "Collection not initialized"
```

**2. Bidirectional relationships — define both sides**:
```typescript
@Entity()
class Book {
  @ManyToOne(() => Author, { nullable: false })
  author: Author;  // owning side
}

@Entity()
class Author {
  @OneToMany(() => Book, book => book.author)
  books = new Collection<Book>(this);  // inverse side
}
```

**3. M:N relationships — use an explicit pivot entity for extra fields**:
```typescript
@Entity()
class BookTag {
  @ManyToOne(() => Book, { primary: true }) book: Book;
  @ManyToOne(() => Tag, { primary: true }) tag: Tag;
  @Property() addedAt = new Date();  // Can add fields to the join
}

// Instead of:
// @ManyToMany(() => Tag)
// tags = new Collection<Tag>(this);  // No room for extra fields
```

**4. The EntityManager pattern — always flush explicitly**:
```typescript
// Don't forget to flush — changes are tracked in memory until flushed
author.name = "Updated Name";
await em.persistAndFlush(author);  // persist + flush in one call
// or:
em.persist(author);
await em.flush();  // Flushes all pending changes
```

What specific relationship pattern is causing issues? M:N bidirectional with MongoDB has some specific constraints compared to SQL backends.
